File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-go/llm/fake_chat_model.go
package llm

import (
	"errors"
	"sync"

	"opspilot/tools"
)

// Responder simula comportamento dependente do input, equivalente ao
// `Function<List<ChatMessage>, ModelResponse>` do porte Java.
type Responder func(messages []ChatMessage) (ModelResponse, error)

// FakeChatModel é o test double sem chamada de rede. O UNIDADE.md da unidade
// 9 registra que "o código passa no typecheck e nos testes com fakes" --
// este é o equivalente Go desses fakes: uma fila de respostas programadas, ou
// uma função arbitrária.
type FakeChatModel struct {
	mu          sync.Mutex
	scripted    []ModelResponse
	responder   Responder
	callHistory [][]ChatMessage
}

var _ ChatModel = (*FakeChatModel)(nil)

// NewFakeChatModelScripted cria um FakeChatModel que consome a fila de
// respostas programadas em ordem, uma por chamada.
func NewFakeChatModelScripted(responses []ModelResponse) *FakeChatModel {
	scripted := make([]ModelResponse, len(responses))
	copy(scripted, responses)
	return &FakeChatModel{scripted: scripted}
}

// NewFakeChatModelFunc cria um FakeChatModel cuja resposta é calculada por
// uma função arbitrária a partir das mensagens recebidas.
func NewFakeChatModelFunc(responder Responder) *FakeChatModel {
	return &FakeChatModel{responder: responder}
}

func (f *FakeChatModel) Invoke(messages []ChatMessage, _ []tools.Tool) (ModelResponse, error) {
	f.mu.Lock()
	defer f.mu.Unlock()

	f.callHistory = append(f.callHistory, messages)

	if f.responder != nil {
		return f.responder(messages)
	}
	if len(f.scripted) == 0 {
		return ModelResponse{}, errors.New("FakeChatModel ran out of scripted responses")
	}
	next := f.scripted[0]
	f.scripted = f.scripted[1:]
	return next, nil
}

// CallCount é o porte de FakeChatModel.callCount() (porte Java).
func (f *FakeChatModel) CallCount() int {
	f.mu.Lock()
	defer f.mu.Unlock()
	return len(f.callHistory)
}

File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-java/src/main/java/com/opspilot/llm/FakeChatModel.java
package com.opspilot.llm;

import com.opspilot.tools.Tool;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import java.util.function.Function;

/**
 * Test double sem chamada de rede. O UNIDADE.md da unidade 9 registra que
 * "o codigo passa no typecheck e nos testes com fakes" -- este e o
 * equivalente Java desses fakes: uma fila de respostas programadas, ou uma
 * funcao arbitraria (para simular comportamento dependente do input).
 */
public class FakeChatModel implements ChatModel {

    private final Deque<ModelResponse> scriptedResponses;
    private final Function<List<ChatMessage>, ModelResponse> responder;
    private final List<List<ChatMessage>> callHistory = new ArrayList<>();

    public FakeChatModel(List<ModelResponse> scriptedResponses) {
        this.scriptedResponses = new ArrayDeque<>(scriptedResponses);
        this.responder = null;
    }

    public FakeChatModel(Function<List<ChatMessage>, ModelResponse> responder) {
        this.scriptedResponses = null;
        this.responder = responder;
    }

    @Override
    public ModelResponse invoke(List<ChatMessage> messages, List<Tool> tools) {
        callHistory.add(messages);
        if (responder != null) {
            return responder.apply(messages);
        }
        if (scriptedResponses.isEmpty()) {
            throw new IllegalStateException("FakeChatModel ran out of scripted responses");
        }
        return scriptedResponses.poll();
    }

    public int callCount() {
        return callHistory.size();
    }
}
